platforms/allwinner-d1/core/src/drivers/twi.rs

```diff
@@ -0,0 +1,809 @@
+        // reset the TWI engine.
+        guard.twi.twi_srst.write(|w| w.soft_rst().set_bit());
+        guard.twi.twi_efr.reset();
+        guard.twi.twi_data.reset();
+
+        let mut started = false;
+        while let Ok(Transfer {
+            buf,
+            len,
+            end,
+            dir,
+            rsp,
+        }) = txn.dequeue_async().await
+        {
+            // setup TWI driver state for next op
+            guard.data.state = if started {
+                State::WaitForRestart(addr)
+            } else {
+                started = true;
+                State::WaitForStart(addr)
+            };
+            guard.data.op = match dir {
+                OpKind::Read => {
+                    // setup read op
+                    tracing::debug!("reading {len} bytes");
+                    TwiOp::Read {
+                        buf,
+                        len,
+                        amt: 0,
+                        end,
+                    }
+                }
+                OpKind::Write => {
+                    // setup write op
+                    tracing::debug!("writing {len} bytes");
+                    TwiOp::Write {
+                        buf,
+                        pos: 0,
+                        len,
+                        end,
+                    }
+                }
+            };
+
+            guard.wait_for_irq().await;
+            tracing::trace!(?dir, "TWI operation completed");
+            let res = if let Some(error) = guard.data.err.take() {
+                tracing::warn!(?error, ?dir, "TWI error");
+                Err(error)
+            } else {
+                match core::mem::replace(&mut guard.data.op, TwiOp::None) {
+                    TwiOp::Read { buf, .. } => {
+                        debug_assert_eq!(dir, OpKind::Read);
+                        Ok(buf)
+                    }
+                    TwiOp::Write { buf, .. } => {
+                        debug_assert_eq!(dir, OpKind::Write);
+                        Ok(buf)
+                    }
+                    _ => unreachable!(),
+                }
+            };
+            if rsp.send(res).is_err() {
+                tracing::trace!("I2C transaction handle dropped");
+                break;
+            }
+        }
+        // transaction ended!
+        tracing::debug!("I2C transaction ended");
+
+        let guard = self.isr.lock(self.twi);
+        guard.twi.twi_cntr.modify(|_r, w| {
+            w.m_stp().set_bit();
+            w
+        });
+    }
+}
+
+impl IsrData {
+    #[must_use]
+    fn lock<'a>(&'a self, twi: &'a twi::RegisterBlock) -> TwiDataGuard<'a> {
+        // disable TWI interrupts while holding the guard.
+        twi.twi_cntr.modify(|_r, w| w.int_en().low());
+        let data = unsafe { &mut *(self.data.get()) };
+        TwiDataGuard { data, twi }
+    }
+}
+
+impl Drop for TwiDataGuard<'_> {
+    fn drop(&mut self) {
+        self.twi.twi_cntr.modify(|_r, w| w.int_en().high());
```

Comment:
Should we check if the int *wasn't* enabled before locking? not sure if this matters, but locking and unlocking could indirectly enable the interrupt even if you weren't expecting it.

Reply:
theoretically yeah, we probably should — its not actually an issue here because the guard type is only used in one place, but it would be more correct to do that.

Reply:
really, the `MutexGuard`-like behavior is not actually necessary any longer. at one point, i was actually locking and dropping the guard in several places. now, the only time we actually unmask the interrupt is inside of `wait_for_irq`, which guarantees that the driver task is not writing to the shared state just because it's yielded to wait for the IRQ. so, we could refactor this to remove the guard-like behavior...